wgsl/index.bs

```diff
@@ -4557,11 +4555,282 @@ TODO: *Stub*: Expression evaluation
+    * Else remember it for later.
+* Look at which vertices are reachable from "True"
+    * If this set includes the vertex "False", then reject the program.
+    * Remember for later whether this set includes "CF_start" and/or any of the "arg_i"
+    * (optional, makes the analysis faster) Remove from the graph all vertices that have been visited
+* Look at which vertices are reachable from "CF_return"
+    * Remember for later whether this set includes "False" and/or any of the "arg_i"
+    * (optional, makes the analysis faster) Remove from the graph all vertices that have been visited
+* If "Value_return" exists, look at which vertices are reachable from it
+    * Remember for later whether this set includes "False" and/or any of the "arg_i"
+
+Note: Anything which is not explicitly remembered, including the entire graph can be destroyed at this point. The little information that we do need to remember to analyze the caller is linear in the number of parameters of the function.
+
+### Uniformity rules for statements ### {#uniformity-statements}
 
-### Divergence and reconvergence TODO ### {#divergence-reconvergence}
+In order to properly deal with reconvergence we must know when there is a single way for control to flow out of a statement.
+We achieve this by assigning to each statement a set of *behaviors* out of {Return, Break, Continue, Discard, Nothing}.
+Intuitively, this set corresponds to the ways that control-flow can exit the statement.
 
-### Uniformity restrictions TODO ### {#uniformity-restrictions}
+TODO: do we need an extra behavior for possibly infinite loops? I should check whether some form of reconvergence happens after something like `while(tid mod 2) {}`.
+
+If the set of behaviors for a statement has a single element, this means that we either did not diverge within the statement, or we reconverged, and we pick the vertex corresponding to control-flow at the start of the statement as the vertex corresponding to control-flow at the exit of the statement, overriding the other rules below.
+
+The rules for analyzing statements take as argument both the statement itself and the vertex corresponding to control-flow at the beginning of it (which we'll note "CF" below) and return the following:
+
+* A set of behaviors for the statement
+* A vertex corresponding to control-flow at the exit of it
+* A set of new vertices and edges to add to the graph
+
+In the table below, `CF1, S -> CF2, B` means "run the analysis on S starting with control-flow CF1, apply the required changes to the graph, and name the resulting control-flow CF2, and the resulting set of behaviors B".
+Similarly, `CF, E -> CF, V, B` means "run the analysis on expression E, starting with control-flow CF, apply the required changes to the graph, and name the resulting control-flow vertex CF, the resulting value vertex V, and the resulting set of behaviors B" (see next section for the analysis of expressions).
```

Comment:
I don't understand this part. We are starting with control-flow CF. Does it have a vertex in the graph at this point?
Why do we "name the resulting control-flow vertex CF" then?

Reply:
Sorry, that my fault for having a typo, the second CF should be CF2. Fixing it.